// typeorm-spatial/build/compiled/test/functional/spatial/postgres/spatial-postgres.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var _this = this;
Object.defineProperty(exports, "__esModule", { value: true });
require("reflect-metadata");
var chai_1 = require("chai");
var test_utils_1 = require("../../../utils/test-utils");
var Post_1 = require("./entity/Post");
describe("spatial-postgres", function () {
    var connections;
    before(function () { return __awaiter(_this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, test_utils_1.createTestingConnections({
                        entities: [__dirname + "/entity/*{.js,.ts}"],
                        enabledDrivers: ["postgres"]
                    })];
                case 1:
                    connections = _a.sent();
                    return [2 /*return*/];
            }
        });
    }); });
    beforeEach(function () { return __awaiter(_this, void 0, void 0, function () {
        var err_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, test_utils_1.reloadTestingDatabases(connections)];
                case 1:
                    _a.sent();
                    return [3 /*break*/, 3];
                case 2:
                    err_1 = _a.sent();
                    console.warn(err_1.stack);
                    throw err_1;
                case 3: return [2 /*return*/];
            }
        });
    }); });
    after(function () { return __awaiter(_this, void 0, void 0, function () {
        var err_2;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, test_utils_1.closeTestingConnections(connections)];
                case 1:
                    _a.sent();
                    return [3 /*break*/, 3];
                case 2:
                    err_2 = _a.sent();
                    console.warn(err_2.stack);
                    throw err_2;
                case 3: return [2 /*return*/];
            }
        });
    }); });
    it("should create correct schema with Postgres' geometry type", function () {
        return Promise.all(connections.map(function (connection) { return __awaiter(_this, void 0, void 0, function () {
            var queryRunner, schema;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        queryRunner = connection.createQueryRunner();
                        return [4 /*yield*/, queryRunner.getTable("post")];
                    case 1:
                        schema = _a.sent();
                        return [4 /*yield*/, queryRunner.release()];
                    case 2:
                        _a.sent();
                        chai_1.expect(schema).not.to.be.empty;
                        chai_1.expect(schema.columns.find(function (tableColumn) {
                            return tableColumn.name === "geom" && tableColumn.type === "geometry";
                        })).to.not.be.empty;
                        return [2 /*return*/];
                }
            });
        }); }));
    });
    it("should create correct schema with Postgres' geography type", function () {
        return Promise.all(connections.map(function (connection) { return __awaiter(_this, void 0, void 0, function () {
            var queryRunner, schema;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        queryRunner = connection.createQueryRunner();
                        return [4 /*yield*/, queryRunner.getTable("post")];
                    case 1:
                        schema = _a.sent();
                        return [4 /*yield*/, queryRunner.release()];
                    case 2:
                        _a.sent();
                        chai_1.expect(schema).not.to.be.empty;
                        chai_1.expect(schema.columns.find(function (tableColumn) {
                            return tableColumn.name === "geog" && tableColumn.type === "geography";
                        })).to.not.be.empty;
                        return [2 /*return*/];
                }
            });
        }); }));
    });
    it("should create correct schema with Postgres' geometry indices", function () {
        return Promise.all(connections.map(function (connection) { return __awaiter(_this, void 0, void 0, function () {
            var queryRunner, schema;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        queryRunner = connection.createQueryRunner();
                        return [4 /*yield*/, queryRunner.getTable("post")];
                    case 1:
                        schema = _a.sent();
                        return [4 /*yield*/, queryRunner.release()];
                    case 2:
                        _a.sent();
                        chai_1.expect(schema).not.to.be.empty;
                        chai_1.expect(schema.indices.find(function (tableIndex) {
                            return tableIndex.isSpatial === true &&
                                tableIndex.columnNames.length === 1 &&
                                tableIndex.columnNames[0] === "geom";
                        })).to.not.be.empty;
                        return [2 /*return*/];
                }
            });
        }); }));
    });
    it("should persist geometry correctly", function () {
        return Promise.all(connections.map(function (connection) { return __awaiter(_this, void 0, void 0, function () {
            var geom, recordRepo, post, persistedPost, foundPost;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        geom = {
                            type: "Point",
                            coordinates: [0, 0]
                        };
                        recordRepo = connection.getRepository(Post_1.Post);
                        post = new Post_1.Post();
                        post.geom = geom;
                        return [4 /*yield*/, recordRepo.save(post)];
                    case 1:
                        persistedPost = _a.sent();
                        return [4 /*yield*/, recordRepo.findOne(persistedPost.id)];
                    case 2:
                        foundPost = _a.sent();
                        chai_1.expect(foundPost).to.exist;
                        chai_1.expect(foundPost.geom).to.deep.equal(geom);
                        return [2 /*return*/];
                }
            });
        }); }));
    });
    it("should persist geography correctly", function () {
        return Promise.all(connections.map(function (connection) { return __awaiter(_this, void 0, void 0, function () {
            var geom, recordRepo, post, persistedPost, foundPost;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        geom = {
                            type: "Point",
                            coordinates: [0, 0]
                        };
                        recordRepo = connection.getRepository(Post_1.Post);
                        post = new Post_1.Post();
                        post.geog = geom;
                        return [4 /*yield*/, recordRepo.save(post)];
                    case 1:
                        persistedPost = _a.sent();
                        return [4 /*yield*/, recordRepo.findOne(persistedPost.id)];
                    case 2:
                        foundPost = _a.sent();
                        chai_1.expect(foundPost).to.exist;
                        chai_1.expect(foundPost.geog).to.deep.equal(geom);
                        return [2 /*return*/];
                }
            });
        }); }));
    });
});
